src/main/scala/com.snowplowanalytics.sauna/responders/hipchat/SendRoomNotificationResponder.scala

```diff
@@ -0,0 +1,238 @@
+   * @param command  Specifies the exact action to execute.
+   */
+  case class SaunaCommand(
+    envelope: SelfDescribing,
+    command: SelfDescribing)
+
+  implicit val saunaCommandReads: Reads[SaunaCommand] = (
+    (JsPath \ "envelope").read[SelfDescribing] and
+      (JsPath \ "command").read[SelfDescribing]
+    ) (SaunaCommand.apply _)
+
+  /**
+   * TODO: define what this is
+   */
+  object Semantics extends Enumeration {
+    type Semantics = Value
+    val AT_LEAST_ONCE = Value("AT_LEAST_ONCE")
+  }
+
+  implicit val semanticsReads: Reads[Semantics] = Reads.enumNameReads(Semantics)
+
+  /**
+   * TODO: define what this is
+   *
+   * @param semantics  TODO: define what this is
+   * @param timeToLive The command's lifetime (in ms): if a command is
+   *                   received after `timeToLive` ms from its' creation have
+   *                   passed, it will be discarded. (If None, a command won't be
+   *                   discarded based on creation time.)
+   */
+  case class ExecutionParams(
+    semantics: Semantics,
+    timeToLive: Option[Int]
+  )
+
+  implicit val timeToLiveReads: Reads[Option[Int]] = JsPath.readNullable[Int]
+  implicit val executionParamsReads: Reads[ExecutionParams] = Json.reads[ExecutionParams]
+
+  /**
+   * A Sauna command envelope.
+   *
+   * @param commandId   A unique identifier (uuid4) for the command.
+   * @param whenCreated The command's creation date.
+   * @param execution   TODO: define what this is
+   * @param tags        TODO: define what this is
+   */
+  case class CommandEnvelope(
+    commandId: String,
+    whenCreated: LocalDateTime,
+    execution: ExecutionParams,
+    tags: Map[String, String]
+  )
+  implicit val commandEnvelopeReads: Reads[CommandEnvelope] = Json.reads[CommandEnvelope]
+
+  /**
+   * Constructs a [[Props]] for a [[SendRoomNotificationResponder]] actor.
+   *
+   * @param hipchat The HipChat API wrapper.
+   * @param logger  A logger actor.
+   * @return [[Props]] for the new actor.
+   */
+  def props(hipchat: Hipchat, logger: ActorRef): Props =
+    Props(new SendRoomNotificationResponder(hipchat, logger))
+
+  /**
+   * Attempts to extract a Sauna command from a [[JsValue]].
+   *
+   * @param json The [[JsValue]] to extract a command from.
+   * @tparam T The type of the command's data.
+   * @return Right containing a tuple of the command's envelope and data
+   *         if the extraction was successful, Left containing an error message
+   *         otherwise.
+   */
+  def extractCommand[T](json: JsValue)(implicit tReads: Reads[T]): Either[String, (CommandEnvelope, T)] = {
+    json.validate[SelfDescribing] match {
+      case JsSuccess(selfDescribing, _) =>
+        selfDescribing.data.validate[SaunaCommand] match {
+          case JsSuccess(command, _) =>
+            command.envelope.data.validate[CommandEnvelope] match {
+              case JsSuccess(envelope, _) =>
+                command.command.data.validate[T] match {
```

Comment:
Hm. I'm quite surprised `validate[T]` works as expected in test. From my perspective type-erasure should prevent this from proper working.